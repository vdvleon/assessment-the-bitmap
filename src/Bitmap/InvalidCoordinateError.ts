
export default class InvalidCoordinateError extends Error {
    constructor(x: number, y: number, width: number, height: number) {
        super(`Invalid coordinate given: (${x}, ${y}) for bitmap of size ${width} x ${height}`);
    }
}
