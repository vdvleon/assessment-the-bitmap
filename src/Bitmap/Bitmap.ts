import InvalidCoordinateError from './InvalidCoordinateError';

export default class Bitmap<T> {
    width: number;
    height: number;
    data: T[][];

    /**
     * Initialize a bitmap for a given size.
     *
     * @param width
     * @param height
     * @param initialValue
     */
    constructor(width: number, height: number, initialValue: T) {
        this.width = width;
        this.height = height;
        this.data = [];

        // Initialize the 'matrix'
        for (let h = 0; h < height; ++h) {
            this.data[h] = [];
            for (let w = 0; w < width; ++w) {
                this.data[h][w] = initialValue;
            }
        }
    }

    setValueAt(x: number, y: number, value: T): this {
        this.validateCoordinate(x, y);

        this.data[y][x] = value;

        return this;
    }

    getValueAt(x: number, y: number): T {
        this.validateCoordinate(x, y);

        return this.data[y][x];
    }

    validateCoordinate(x, y) {
        if (x < 0 || x >= this.width || y < 0 || y >= this.height) {
            throw new InvalidCoordinateError(x, y, this.width, this.height);
        }
    }
}
