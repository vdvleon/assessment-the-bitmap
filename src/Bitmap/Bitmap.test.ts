import Bitmap from './Bitmap';

describe('Bitmap', () => {
    describe('constructor', () => {
        it('should create a bitmap with a data matrix for given size', () => {
            const width = 2;
            const height = 3;
            const bitmap = new Bitmap<boolean>(width, height, false);

            expect(bitmap.data).toEqual(height);
            for (let h = 0; h < height; ++h) {
                expect(bitmap.data[h].length).toEqual(width);
            }
        });

        it('should create a bitmap with a data matrix of given type', () => {
            const bitmap1 = new Bitmap<string>(1, 1, 'foobar');
            expect(typeof bitmap1.data[0][0]).toBe('string');

            const bitmap2 = new Bitmap<number>(1, 1, 13.37);
            expect(typeof bitmap2.data[0][0]).toBe('number');

            const bitmap3 = new Bitmap<Date>(1, 1, new Date(0));
            expect(bitmap3.data[0][0]).toBeInstanceOf(Date);
        });
    });

    describe('setValueAt', () => {
       it('should fail when given an incorrect coordinate', () => {
          const bitmap = new Bitmap<boolean>(1, 1, false);
          expect(() => bitmap.setValueAt(1, 1, true)).toThrow(Error);
       });

        it('should set value for a valid coordinate and return this', () => {
            const bitmap = new Bitmap<boolean>(1, 1, false);

            expect(bitmap.setValueAt(0, 0, true)).toBe(bitmap);
            expect(bitmap.data[0][0]).toBe(true);
        });
    });

    describe('getValueAt', () => {
        it('should fail when given an incorrect coordinate', () => {
            const bitmap = new Bitmap<boolean>(1, 1, false);
            expect(() => bitmap.getValueAt(1, 1)).toThrow(Error);
        });

        it('should get value for a valid coordinate', () => {
            const bitmap = new Bitmap<boolean>(1, 1, false);

            expect(bitmap.getValueAt(0, 0)).toBe(false);

            bitmap.setValueAt(0, 0, true);

            expect(bitmap.getValueAt(0, 0)).toBe(true);
        });
    });
});
