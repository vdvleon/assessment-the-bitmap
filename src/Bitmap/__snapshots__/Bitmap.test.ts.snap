// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bitmap constructor should create a bitmap with a data matrix for given size 1`] = `
Array [
  Array [
    false,
    false,
  ],
  Array [
    false,
    false,
  ],
  Array [
    false,
    false,
  ],
]
`;
